lib/src/collecting_body.rs

```diff
@@ -0,0 +1,340 @@
+//! Provides `CollectingBody`, a body that can be concurrently written-to (via a `StreamingBody`
+//! handle) and read-from (via multiple `Body`) handles.
+
+// TODO: cceckman-at-fastly: Temporary, I promise.
+#![allow(unused)]
+
+use bytes::Bytes;
+use http::HeaderMap;
+use http_body::Body as HttpBody;
+use tokio::sync::watch;
+
+use crate::{
+    body::{Body, Chunk},
+    error,
+    streaming_body::StreamingBody,
+    Error,
+};
+
+/// CollectingBody is a body for caching and request collapsing.
+/// It allows writing a body while concurrently\* reading it, from multiple readers.
+#[derive(Debug)]
+pub struct CollectingBody {
+    inner: watch::Receiver<CollectedBody>,
+}
+
+/// Body consisting strictly of Byte chunks, with concurrent readying and writing.
+///
+/// CollectingBody primarily exists to implement the cache APIs. Cache APIs allow three things to
+/// happen concurrently:
+/// - Streaming in a body, from an `insert`/`replace`-style request, implemented via StreamingBody
+/// - Storage of that body in the cache
+/// - Streaming out of the body, from a `lookup`/`TransactionInsertBuilder:execute_and_stream_back`
+///   request, from the same or a different session
+///
+/// CollectingBody provides a place for this to happen. It accepts a `Body` as a source of data,
+/// e.g. one from a `StreamingBody` or an origin response; stores the data for future retrieval;
+/// and can return new `Body`s from `::read` that produce the full content.
+///
+/// A CollectingBody is finished when either error.is_some() or trailers.is_some().
+/// Until then, more chunks may be added!
+#[derive(Default, Debug)]
+struct CollectedBody {
+    // We preserve "multiple Byte chunks" here to avoid additional copies-
+    // the Byte type uses a shared buffer under the hood.
+    // TODO: cceckman-at-fastly: consider SmallVec
+    chunks: Vec<Bytes>,
+    error: Option<Error>,
+    trailers: Option<HeaderMap>,
+}
+
+impl CollectingBody {
```

Comment:
Since `CollectingBody` and `CollectedBody` are very similar names can you swap the `struct` ordering so the `struct` and `impl` for `CollectingBody` are adjacent? It gave me a moment of confusion when reading the impl.